File: vox/swift/roam-runtime/Sources/RoamRuntime/Schema.swift
import Foundation

public typealias SchemaHash = UInt64
public typealias TypeParamName = String
public typealias SchemaRegistry = [SchemaHash: Schema]

public indirect enum TypeRef: Sendable, Equatable {
    case concrete(typeId: SchemaHash, args: [TypeRef])
    case `var`(name: TypeParamName)

    public static func concrete(_ typeId: SchemaHash) -> Self {
        .concrete(typeId: typeId, args: [])
    }

    public static func generic(_ typeId: SchemaHash, args: [TypeRef]) -> Self {
        .concrete(typeId: typeId, args: args)
    }

    func encode() -> [UInt8] {
        switch self {
        case .concrete(let typeId, let args):
            return encodeVarint(0) + encodeVarint(typeId) + encodeVec(args, encoder: { $0.encode() })
        case .var(let name):
            return encodeVarint(1) + encodeString(name)
        }
    }

    static func decode(from data: Data, offset: inout Int) throws -> Self {
        let disc = try decodeVarint(from: data, offset: &offset)
        switch disc {
        case 0:
            let typeId = try decodeVarint(from: data, offset: &offset)
            let args = try decodeVec(
                from: data,
                offset: &offset,
                decoder: { data, off in try TypeRef.decode(from: data, offset: &off) }
            )
            return .concrete(typeId: typeId, args: args)
        case 1:
            return .var(name: try decodeString(from: data, offset: &offset))
        default:
            throw PostcardError.unknownVariant
        }
    }
}

public struct Schema: Sendable, Equatable {
    public var id: SchemaHash
    public var typeParams: [TypeParamName]
    public var kind: SchemaKind

    public init(id: SchemaHash, typeParams: [TypeParamName], kind: SchemaKind) {
        self.id = id
        self.typeParams = typeParams
        self.kind = kind
    }

    func encode() -> [UInt8] {
        encodeVarint(id)
            + encodeVec(typeParams, encoder: encodeString)
            + kind.encode()
    }

    static func decode(from data: Data, offset: inout Int) throws -> Self {
        let id = try decodeVarint(from: data, offset: &offset)
        let typeParams = try decodeVec(from: data, offset: &offset, decoder: decodeString)
        let kind = try SchemaKind.decode(from: data, offset: &offset)
        return .init(id: id, typeParams: typeParams, kind: kind)
    }

    public var name: String? {
        switch kind {
        case .struct(let name, _), .enum(let name, _):
            return name
        default:
            return nil
        }
    }
}

public indirect enum SchemaKind: Sendable, Equatable {
    case `struct`(name: String, fields: [FieldSchema])
    case `enum`(name: String, variants: [VariantSchema])
    case tuple(elements: [TypeRef])
    case list(element: TypeRef)
    case map(key: TypeRef, value: TypeRef)
    case array(element: TypeRef, length: UInt64)
    case option(element: TypeRef)
    case channel(direction: ChannelDirection, element: TypeRef)
    case primitive(PrimitiveType)

    func encode() -> [UInt8] {
        switch self {
        case .struct(let name, let fields):
            return encodeVarint(0) + encodeString(name) + encodeVec(fields, encoder: { $0.encode() })
        case .enum(let name, let variants):
            return encodeVarint(1) + encodeString(name) + encodeVec(variants, encoder: { $0.encode() })
        case .tuple(let elements):
            return encodeVarint(2) + encodeVec(elements, encoder: { $0.encode() })
        case .list(let element):
            return encodeVarint(3) + element.encode()
        case .map(let key, let value):
            return encodeVarint(4) + key.encode() + value.encode()
        case .array(let element, let length):
            return encodeVarint(5) + element.encode() + encodeVarint(length)
        case .option(let element):
            return encodeVarint(6) + element.encode()
        case .channel(let direction, let element):
            return encodeVarint(7) + direction.encode() + element.encode()
        case .primitive(let primitiveType):
            return encodeVarint(8) + primitiveType.encode()
        }
    }

    static func decode(from data: Data, offset: inout Int) throws -> Self {
        let disc = try decodeVarint(from: data, offset: &offset)
        switch disc {
        case 0:
            return .struct(
                name: try decodeString(from: data, offset: &offset),
                fields: try decodeVec(
                    from: data,
                    offset: &offset,
                    decoder: { data, off in try FieldSchema.decode(from: data, offset: &off) }
                )
            )
        case 1:
            return .enum(
                name: try decodeString(from: data, offset: &offset),
                variants: try decodeVec(
                    from: data,
                    offset: &offset,
                    decoder: { data, off in try VariantSchema.decode(from: data, offset: &off) }
                )
            )
        case 2:
            return .tuple(
                elements: try decodeVec(
                    from: data,
                    offset: &offset,
                    decoder: { data, off in try TypeRef.decode(from: data, offset: &off) }
                )
            )
        case 3:
            return .list(element: try TypeRef.decode(from: data, offset: &offset))
        case 4:
            return .map(
                key: try TypeRef.decode(from: data, offset: &offset),
                value: try TypeRef.decode(from: data, offset: &offset)
            )
        case 5:
            return .array(
                element: try TypeRef.decode(from: data, offset: &offset),
                length: try decodeVarint(from: data, offset: &offset)
            )
        case 6:
            return .option(element: try TypeRef.decode(from: data, offset: &offset))
        case 7:
            return .channel(
                direction: try ChannelDirection.decode(from: data, offset: &offset),
                element: try TypeRef.decode(from: data, offset: &offset)
            )
        case 8:
            return .primitive(try PrimitiveType.decode(from: data, offset: &offset))
        default:
            throw PostcardError.unknownVariant
        }
    }
}

public enum ChannelDirection: Sendable, Equatable {
    case tx
    case rx

    func encode() -> [UInt8] {
        switch self {
        case .tx:
            return encodeVarint(0)
        case .rx:
            return encodeVarint(1)
        }
    }

    static func decode(from data: Data, offset: inout Int) throws -> Self {
        switch try decodeVarint(from: data, offset: &offset) {
        case 0:
            return .tx
        case 1:
            return .rx
        default:
            throw PostcardError.unknownVariant
        }
    }
}

public struct FieldSchema: Sendable, Equatable {
    public var name: String
    public var typeRef: TypeRef
    public var required: Bool

    public init(name: String, typeRef: TypeRef, required: Bool) {
        self.name = name
        self.typeRef = typeRef
        self.required = required
    }

    func encode() -> [UInt8] {
        encodeString(name) + typeRef.encode() + encodeBool(required)
    }

    static func decode(from data: Data, offset: inout Int) throws -> Self {
        .init(
            name: try decodeString(from: data, offset: &offset),
            typeRef: try TypeRef.decode(from: data, offset: &offset),
            required: try decodeBool(from: data, offset: &offset)
        )
    }
}

public struct VariantSchema: Sendable, Equatable {
    public var name: String
    public var index: UInt32
    public var payload: VariantPayload

    public init(name: String, index: UInt32, payload: VariantPayload) {
        self.name = name
        self.index = index
        self.payload = payload
    }

    func encode() -> [UInt8] {
        encodeString(name) + encodeVarint(UInt64(index)) + payload.encode()
    }

    static func decode(from data: Data, offset: inout Int) throws -> Self {
        let name = try decodeString(from: data, offset: &offset)
        let rawIndex = try decodeVarint(from: data, offset: &offset)
        guard rawIndex <= UInt64(UInt32.max) else {
            throw PostcardError.overflow
        }
        return .init(
            name: name,
            index: UInt32(rawIndex),
            payload: try VariantPayload.decode(from: data, offset: &offset)
        )
    }
}

public indirect enum VariantPayload: Sendable, Equatable {
    case unit
    case newtype(typeRef: TypeRef)
    case tuple(types: [TypeRef])
    case `struct`(fields: [FieldSchema])

    func encode() -> [UInt8] {
        switch self {
        case .unit:
            return encodeVarint(0)
        case .newtype(let typeRef):
            return encodeVarint(1) + typeRef.encode()
        case .tuple(let types):
            return encodeVarint(2) + encodeVec(types, encoder: { $0.encode() })
        case .struct(let fields):
            return encodeVarint(3) + encodeVec(fields, encoder: { $0.encode() })
        }
    }

    static func decode(from data: Data, offset: inout Int) throws -> Self {
        let disc = try decodeVarint(from: data, offset: &offset)
        switch disc {
        case 0:
            return .unit
        case 1:
            return .newtype(typeRef: try TypeRef.decode(from: data, offset: &offset))
        case 2:
            return .tuple(
                types: try decodeVec(
                    from: data,
                    offset: &offset,
                    decoder: { data, off in try TypeRef.decode(from: data, offset: &off) }
                )
            )
        case 3:
            return .struct(
                fields: try decodeVec(
                    from: data,
                    offset: &offset,
                    decoder: { data, off in try FieldSchema.decode(from: data, offset: &off) }
                )
            )
        default:
            throw PostcardError.unknownVariant
        }
    }
}

public enum PrimitiveType: Sendable, Equatable {
    case bool
    case u8
    case u16
    case u32
    case u64
    case u128
    case i8
    case i16
    case i32
    case i64
    case i128
    case f32
    case f64
    case char
    case string
    case unit
    case never
    case bytes
    case payload

    func encode() -> [UInt8] {
        let disc: UInt64 = switch self {
        case .bool: 0
        case .u8: 1
        case .u16: 2
        case .u32: 3
        case .u64: 4
        case .u128: 5
        case .i8: 6
        case .i16: 7
        case .i32: 8
        case .i64: 9
        case .i128: 10
        case .f32: 11
        case .f64: 12
        case .char: 13
        case .string: 14
        case .unit: 15
        case .never: 16
        case .bytes: 17
        case .payload: 18
        }
        return encodeVarint(disc)
    }

    static func decode(from data: Data, offset: inout Int) throws -> Self {
        switch try decodeVarint(from: data, offset: &offset) {
        case 0: return .bool
        case 1: return .u8
        case 2: return .u16
        case 3: return .u32
        case 4: return .u64
        case 5: return .u128
        case 6: return .i8
        case 7: return .i16
        case 8: return .i32
        case 9: return .i64
        case 10: return .i128
        case 11: return .f32
        case 12: return .f64
        case 13: return .char
        case 14: return .string
        case 15: return .unit
        case 16: return .never
        case 17: return .bytes
        case 18: return .payload
        default:
            throw PostcardError.unknownVariant
        }
    }
}

public struct SchemaPayload: Sendable, Equatable {
    public var schemas: [Schema]
    public var root: TypeRef

    public init(schemas: [Schema], root: TypeRef) {
        self.schemas = schemas
        self.root = root
    }
}

public enum BindingDirection: Sendable, Equatable {
    case args
    case response
}

public struct SchemaSet: Sendable, Equatable {
    public var root: Schema
    public var registry: SchemaRegistry

    public init(root: Schema, registry: SchemaRegistry) {
        self.root = root
        self.registry = registry
    }
}

public func buildSchemaRegistry(_ schemas: [Schema]) -> SchemaRegistry {
    var registry: SchemaRegistry = [:]
    for schema in schemas {
        registry[schema.id] = schema
    }
    return registry
}

public func resolveTypeRef(_ ref: TypeRef, in registry: SchemaRegistry) -> SchemaKind? {
    switch ref {
    case .var:
        return nil
    case .concrete(let typeId, let args):
        guard let schema = registry[typeId] else {
            return nil
        }
        guard !args.isEmpty else {
            return schema.kind
        }
        var substitutions: [TypeParamName: TypeRef] = [:]
        for (index, typeParam) in schema.typeParams.enumerated() where index < args.count {
            substitutions[typeParam] = args[index]
        }
        return substituteTypeRefs(in: schema.kind, substitutions: substitutions)
    }
}

private func substituteTypeRef(_ ref: TypeRef, substitutions: [TypeParamName: TypeRef]) -> TypeRef {
    switch ref {
    case .var(let name):
        return substitutions[name] ?? ref
    case .concrete(let typeId, let args):
        return .concrete(typeId: typeId, args: args.map { substituteTypeRef($0, substitutions: substitutions) })
    }
}

private func substituteTypeRefs(in kind: SchemaKind, substitutions: [TypeParamName: TypeRef]) -> SchemaKind {
    let sub = { substituteTypeRef($0, substitutions: substitutions) }

    switch kind {
    case .struct(let name, let fields):
        return .struct(
            name: name,
            fields: fields.map { FieldSchema(name: $0.name, typeRef: sub($0.typeRef), required: $0.required) }
        )
    case .enum(let name, let variants):
        return .enum(
            name: name,
            variants: variants.map {
                VariantSchema(name: $0.name, index: $0.index, payload: substituteVariantPayload($0.payload, substitutions: substitutions))
            }
        )
    case .tuple(let elements):
        return .tuple(elements: elements.map(sub))
    case .list(let element):
        return .list(element: sub(element))
    case .map(let key, let value):
        return .map(key: sub(key), value: sub(value))
    case .array(let element, let length):
        return .array(element: sub(element), length: length)
    case .option(let element):
        return .option(element: sub(element))
    case .channel(let direction, let element):
        return .channel(direction: direction, element: sub(element))
    case .primitive:
        return kind
    }
}

private func substituteVariantPayload(
    _ payload: VariantPayload,
    substitutions: [TypeParamName: TypeRef]
) -> VariantPayload {
    let sub = { substituteTypeRef($0, substitutions: substitutions) }

    switch payload {
    case .unit:
        return .unit
    case .newtype(let typeRef):
        return .newtype(typeRef: sub(typeRef))
    case .tuple(let types):
        return .tuple(types: types.map(sub))
    case .struct(let fields):
        return .struct(
            fields: fields.map { FieldSchema(name: $0.name, typeRef: sub($0.typeRef), required: $0.required) }
        )
    }
}

File: vox/rust/vox-codegen/src/targets/swift/schema.rs
//! Swift schema generation for wire schema exchange.
//!
//! Generates runtime schema information for protocol wire schema exchange
//! using the canonical `Schema`/`SchemaKind`/`TypeRef` representation.
//! Type-level channel binding is driven by the same schema registry.

use facet_core::Facet;
use heck::ToLowerCamelCase;
use vox_types::{ServiceDescriptor, ShapeKind, TypeRef, VoxError, classify_shape, extract_schemas};

/// Generate schema code (wire schemas only — binding uses the same registry).
pub fn generate_schemas(service: &ServiceDescriptor) -> String {
    generate_wire_schemas(service)
}

/// Generate wire schema infrastructure for protocol schema exchange.
///
/// Generates:
/// 1. A global schema registry containing all schemas for all methods (deduplicated)
/// 2. Per-method schema ID lists and root TypeRefs for args and response
///
/// At runtime, the Swift code filters schemas through SchemaSendTracker before encoding.
fn generate_wire_schemas(service: &ServiceDescriptor) -> String {
    use crate::render::hex_u64;
    use std::collections::HashMap;
    use vox_types::{Schema, SchemaHash};

    let service_name = service.service_name.to_lower_camel_case();

    // Extract Result and VoxError schemas once (used for wrapping responses)
    let result_extracted =
        extract_schemas(<Result<bool, u32> as Facet<'static>>::SHAPE).expect("Result schema");
    let result_type_id = match &result_extracted.root {
        TypeRef::Concrete { type_id, .. } => *type_id,
        _ => panic!("Result root should be concrete"),
    };

    let vox_error_extracted =
        extract_schemas(<VoxError<std::convert::Infallible> as Facet<'static>>::SHAPE)
            .expect("VoxError schema");
    let vox_error_type_id = match &vox_error_extracted.root {
        TypeRef::Concrete { type_id, .. } => *type_id,
        _ => panic!("VoxError root should be concrete"),
    };

    // Collect all schemas across all methods into a global registry
    let mut global_schemas: HashMap<SchemaHash, Schema> = HashMap::new();

    // Add Result and VoxError schemas
    for schema in result_extracted.schemas.iter() {
        global_schemas.insert(schema.id, schema.clone());
    }
    for schema in vox_error_extracted.schemas.iter() {
        global_schemas.insert(schema.id, schema.clone());
    }

    // Per-method info: (args_schema_ids, args_root, response_schema_ids, response_root)
    struct MethodSchemaInfo {
        args_schema_ids: Vec<SchemaHash>,
        args_root: TypeRef,
        response_schema_ids: Vec<SchemaHash>,
        response_root: TypeRef,
    }
    let mut method_infos: Vec<(u64, MethodSchemaInfo)> = Vec::new();

    for method in service.methods {
        let method_id = crate::method_id(method);

        // Extract args schemas
        let args_extracted = extract_schemas(method.args_shape).expect("args schema extraction");
        let args_schema_ids: Vec<SchemaHash> =
            args_extracted.schemas.iter().map(|s| s.id).collect();
        for schema in args_extracted.schemas.iter().cloned() {
            global_schemas.insert(schema.id, schema);
        }

        // Extract response schemas - wrap in Result<T, VoxError<E>>
        let (ok_extracted, err_extracted) = match classify_shape(method.return_shape) {
            ShapeKind::Result { ok, err } => (
                extract_schemas(ok).expect("ok schema"),
                extract_schemas(err).expect("err schema"),
            ),
            _ => (
                extract_schemas(method.return_shape).expect("return schema"),
                extract_schemas(<std::convert::Infallible as Facet<'static>>::SHAPE)
                    .expect("Infallible schema"),
            ),
        };

        // Collect response schema IDs (including Result and VoxError)
        let mut response_schema_ids: Vec<SchemaHash> = Vec::new();
        for schema in result_extracted.schemas.iter() {
            response_schema_ids.push(schema.id);
        }
        for schema in vox_error_extracted.schemas.iter() {
            response_schema_ids.push(schema.id);
        }
        for schema in ok_extracted.schemas.iter().cloned() {
            response_schema_ids.push(schema.id);
            global_schemas.insert(schema.id, schema);
        }
        for schema in err_extracted.schemas.iter().cloned() {
            response_schema_ids.push(schema.id);
            global_schemas.insert(schema.id, schema);
        }

        // Deduplicate schema IDs (smaller codegen output)
        let mut seen = std::collections::HashSet::new();
        response_schema_ids.retain(|id| seen.insert(*id));

        // Build the response root: Result<ok_root, VoxError<err_root>>
        let vox_error_ref = TypeRef::generic(vox_error_type_id, vec![err_extracted.root.clone()]);
        let response_root = TypeRef::generic(
            result_type_id,
            vec![ok_extracted.root.clone(), vox_error_ref],
        );

        method_infos.push((
            method_id,
            MethodSchemaInfo {
                args_schema_ids,
                args_root: args_extracted.root.clone(),
                response_schema_ids,
                response_root,
            },
        ));
    }

    let mut out = String::new();

    // Generate global schema registry
    out.push_str("/// Global schema registry containing all schemas for this service.\n");
    out.push_str(&format!(
        "nonisolated(unsafe) public let {service_name}_schema_registry: [UInt64: Schema] = [\n"
    ));

    let mut sorted_schemas: Vec<_> = global_schemas.into_iter().collect();
    sorted_schemas.sort_by_key(|(id, _)| *id);

    for (schema_id, schema) in &sorted_schemas {
        out.push_str(&format!(
            "    {}: {},\n",
            hex_u64(schema_id.0),
            format_swift_schema(schema)
        ));
    }
    out.push_str("]\n\n");

    // Generate per-method schema info
    out.push_str("/// Per-method schema information for wire protocol.\n");
    out.push_str(&format!(
        "nonisolated(unsafe) public let {service_name}_method_schemas: [UInt64: MethodSchemaInfo] = [\n"
    ));

    for (method_id, info) in &method_infos {
        out.push_str(&format!("    {}: MethodSchemaInfo(\n", hex_u64(*method_id)));
        out.push_str(&format!(
            "        argsSchemaIds: [{}],\n",
            info.args_schema_ids
                .iter()
                .map(|id| hex_u64(id.0))
                .collect::<Vec<_>>()
                .join(", ")
        ));
        out.push_str(&format!(
            "        argsRoot: {},\n",
            format_swift_type_ref(&info.args_root)
        ));
        out.push_str(&format!(
            "        responseSchemaIds: [{}],\n",
            info.response_schema_ids
                .iter()
                .map(|id| hex_u64(id.0))
                .collect::<Vec<_>>()
                .join(", ")
        ));
        out.push_str(&format!(
            "        responseRoot: {}\n",
            format_swift_type_ref(&info.response_root)
        ));
        out.push_str("    ),\n");
    }
    out.push_str("]\n\n");

    out
}

/// Format a Schema as Swift code.
fn format_swift_schema(schema: &vox_types::Schema) -> String {
    use crate::render::hex_u64;

    let type_params = if schema.type_params.is_empty() {
        "[]".to_string()
    } else {
        format!(
            "[{}]",
            schema
                .type_params
                .iter()
                .map(|p| format!("\"{}\"", p.as_str()))
                .collect::<Vec<_>>()
                .join(", ")
        )
    };

    format!(
        "Schema(id: {}, typeParams: {}, kind: {})",
        hex_u64(schema.id.0),
        type_params,
        format_swift_schema_kind(&schema.kind)
    )
}

/// Format a SchemaKind as Swift code.
fn format_swift_schema_kind(kind: &vox_types::SchemaKind) -> String {
    use vox_types::SchemaKind;

    match kind {
        SchemaKind::Struct { name, fields } => {
            let fields_str = fields
                .iter()
                .map(|f| {
                    format!(
                        "FieldSchema(name: \"{}\", typeRef: {}, required: {})",
                        f.name,
                        format_swift_type_ref(&f.type_ref),
                        f.required
                    )
                })
                .collect::<Vec<_>>()
                .join(", ");
            format!(".struct(name: \"{}\", fields: [{}])", name, fields_str)
        }
        SchemaKind::Enum { name, variants } => {
            let variants_str = variants
                .iter()
                .map(|v| {
                    format!(
                        "VariantSchema(name: \"{}\", index: {}, payload: {})",
                        v.name,
                        v.index,
                        format_swift_variant_payload(&v.payload)
                    )
                })
                .collect::<Vec<_>>()
                .join(", ");
            format!(".enum(name: \"{}\", variants: [{}])", name, variants_str)
        }
        SchemaKind::Tuple { elements } => {
            let elems_str = elements
                .iter()
                .map(format_swift_type_ref)
                .collect::<Vec<_>>()
                .join(", ");
            format!(".tuple(elements: [{}])", elems_str)
        }
        SchemaKind::List { element } => {
            format!(".list(element: {})", format_swift_type_ref(element))
        }
        SchemaKind::Map { key, value } => {
            format!(
                ".map(key: {}, value: {})",
                format_swift_type_ref(key),
                format_swift_type_ref(value)
            )
        }
        SchemaKind::Array { element, length } => {
            format!(
                ".array(element: {}, length: {})",
                format_swift_type_ref(element),
                length
            )
        }
        SchemaKind::Option { element } => {
            format!(".option(element: {})", format_swift_type_ref(element))
        }
        SchemaKind::Channel { direction, element } => {
            let dir = match direction {
                vox_types::ChannelDirection::Tx => ".tx",
                vox_types::ChannelDirection::Rx => ".rx",
            };
            format!(
                ".channel(direction: {}, element: {})",
                dir,
                format_swift_type_ref(element)
            )
        }
        SchemaKind::Primitive { primitive_type } => {
            format!(".primitive({})", format_swift_primitive(primitive_type))
        }
    }
}

/// Format a VariantPayload as Swift code.
fn format_swift_variant_payload(payload: &vox_types::VariantPayload) -> String {
    use vox_types::VariantPayload;

    match payload {
        VariantPayload::Unit => ".unit".to_string(),
        VariantPayload::Newtype { type_ref } => {
            format!(".newtype(typeRef: {})", format_swift_type_ref(type_ref))
        }
        VariantPayload::Tuple { types } => {
            let types_str = types
                .iter()
                .map(format_swift_type_ref)
                .collect::<Vec<_>>()
                .join(", ");
            format!(".tuple(types: [{}])", types_str)
        }
        VariantPayload::Struct { fields } => {
            let fields_str = fields
                .iter()
                .map(|f| {
                    format!(
                        "FieldSchema(name: \"{}\", typeRef: {}, required: {})",
                        f.name,
                        format_swift_type_ref(&f.type_ref),
                        f.required
                    )
                })
                .collect::<Vec<_>>()
                .join(", ");
            format!(".struct(fields: [{}])", fields_str)
        }
    }
}

/// Format a TypeRef as Swift code.
fn format_swift_type_ref(type_ref: &TypeRef) -> String {
    use crate::render::hex_u64;

    match type_ref {
        TypeRef::Concrete { type_id, args } => {
            if args.is_empty() {
                format!(".concrete({})", hex_u64(type_id.0))
            } else {
                let args_str = args
                    .iter()
                    .map(format_swift_type_ref)
                    .collect::<Vec<_>>()
                    .join(", ");
                format!(".generic({}, args: [{}])", hex_u64(type_id.0), args_str)
            }
        }
        TypeRef::Var { name } => {
            format!(".var(name: \"{}\")", name.as_str())
        }
    }
}

/// Format a PrimitiveType as Swift code.
fn format_swift_primitive(prim: &vox_types::PrimitiveType) -> String {
    use vox_types::PrimitiveType;

    match prim {
        PrimitiveType::Bool => ".bool",
        PrimitiveType::U8 => ".u8",
        PrimitiveType::U16 => ".u16",
        PrimitiveType::U32 => ".u32",
        PrimitiveType::U64 => ".u64",
        PrimitiveType::U128 => ".u128",
        PrimitiveType::I8 => ".i8",
        PrimitiveType::I16 => ".i16",
        PrimitiveType::I32 => ".i32",
        PrimitiveType::I64 => ".i64",
        PrimitiveType::I128 => ".i128",
        PrimitiveType::F32 => ".f32",
        PrimitiveType::F64 => ".f64",
        PrimitiveType::Char => ".char",
        PrimitiveType::String => ".string",
        PrimitiveType::Unit => ".unit",
        PrimitiveType::Never => ".never",
        PrimitiveType::Bytes => ".bytes",
        PrimitiveType::Payload => ".payload",
    }
    .to_string()
}
